tests/test_tomcat.py

```diff
@@ -0,0 +1,118 @@
+""",
+                image_format=ImageFormat.DOCKER,
+            ),
+            marks=marks,
+        )
+    )
+
+
+@pytest.mark.parametrize("container", TOMCAT_CONTAINERS, indirect=True)
+def test_tomcat_launches(container: ContainerData) -> None:
+    """Simple smoke test verifying that the entrypoint launches tomcat and the
+    server is responding to a ``GET /`` on the exposed port.
+
+    """
+
+    # no healthcheck possible, so we have to fallback to get + retry 🫣
+    @tenacity.retry(
+        stop=tenacity.stop_after_attempt(5), wait=tenacity.wait_exponential()
+    )
+    def _fetch_tomcat_root() -> requests.Response:
+        return requests.get(
+            f"http://localhost:{container.forwarded_ports[0].host_port}",
+            timeout=2,
+        )
+
+    resp = _fetch_tomcat_root()
+    baseurl = container.container.baseurl
+    assert baseurl
+    ver = baseurl.rpartition(":")[2]
+    assert resp.status_code == 404
+    assert f"Apache Tomcat/{ver}" in resp.text
+
+
+@pytest.mark.parametrize(
+    "container_per_test", TOMCAT_WITH_SAMPLE, indirect=True
+)
+def test_tomcat_launches_sample_app(container_per_test: ContainerData) -> None:
+    """Launches a container derived from the tomcat containers that includes the
+    `Tomcat sample application
+    <https://tomcat.apache.org/tomcat-10.1-doc/appdev/sample/>`_ in the
+    :file:`$CATALINA_HOME/webapps` directory. Tomcat should serve this sample
+    app automatically via the `GET /sample` route. We test that Tomcat responds
+    to `GET /sample` with a response containing `Sample "Hello, World"
+    Application`.
+
+    """
+    resp = requests.get(
+        f"http://localhost:{container_per_test.forwarded_ports[0].host_port}/sample",
+        timeout=2,
+    )
+    assert resp.status_code == 200
+    assert 'Sample "Hello, World" Application' in resp.text
+
+
+_params = []
+
+for param, pkg in (
+    (TOMCAT_9_CONTAINER, "tomcat"),
+    (TOMCAT_10_CONTAINER, "tomcat10"),
+):
+    ctr, marks = container_and_marks_from_pytest_param(param)
+    _params.append(
+        pytest.param(
+            DerivedContainer(base=ctr, extra_launch_args=["--user", "root"]),
+            pkg,
+            marks=marks,
+        )
+    )
+
+
+@pytest.mark.parametrize(
+    "container_per_test,pkg_name", _params, indirect=["container_per_test"]
+)
+def test_install_samples(
+    container_per_test: ContainerData, pkg_name: str
+) -> None:
+    """Test that we can install the samples application packages for Tomcat 9:
+
+    - ``tomcat-webapps``
+    - ``tomcat-admin-webapps``
+    - ``tomcat-docs-webapp``
+
+    and for Tomcat 10:
+
+    - ``tomcat10-webapps``
+    - ``tomcat10-admin-webapps``
+    - ``tomcat10-docs-webapp``
+
+    """
+    container_per_test.connection.check_output("zypper ref")
+    container_per_test.connection.check_output(
+        f"zypper -n in {pkg_name}-webapps {pkg_name}-admin-webapps {pkg_name}-docs-webapp"
```

Comment:
this package is not in the web and scripting module like "tomcat-admin-webapps" is. why do we include it? this violates the BCI repo policy of being a subset of sles